src/modules/sdbus/sdbus.c

```diff
@@ -0,0 +1,640 @@
+        goto error;
+    }
+    sdbus_recover (ctx, "user requested bus reconnect");
+    if (flux_respond (h, msg, NULL) < 0)
+        flux_log_error (h, "error responding to sdbus.reconnect request");
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, errmsg) < 0)
+        flux_log_error (h, "error responding to sdbus.reconnect request");
+}
+
+static struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST,
+      "sdbus.disconnect",
+      disconnect_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "sdbus.call",
+      call_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "sdbus.subscribe",
+      subscribe_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "sdbus.subscribe-cancel",
+      subscribe_cancel_cb,
+      0
+    },
+    { FLUX_MSGTYPE_REQUEST,
+      "sdbus.reconnect",
+      reconnect_cb,
+      0
+    },
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+/* The bus watcher callback runs sd_bus_process().  Apparently this is an
+ * edge triggered notification so we need to handle all events now, which
+ * means calling sd_bus_process() in a loop until it returns 0.
+ */
+static void sdbus_cb (flux_reactor_t *r,
+                      flux_watcher_t *w,
+                      int revents,
+                      void *arg)
+{
+    struct sdbus_ctx *ctx = arg;
+    int e;
+
+    do {
+        sd_bus_message *m = NULL;
+        if ((e = sd_bus_process (ctx->bus, &m)) < 0) {
+            sdbus_recover (ctx, "error processing sd-bus events");
+            return;
+        }
+        if (m) {
+            // sdbus_recv() may call sdbus_recover() which sets ctx->bus = NULL
+            sdbus_recv (ctx, m);
+            sd_bus_message_unref (m);
+        }
+    } while (e > 0 && ctx->bus != NULL);
+}
+
+/* sdbus.call requests that arrive while the bus connect is in progress
+ * are added to ctx->requests without further processing.  Revisit them now
+ * and begin processing.  Since recovery fails any pending requests, all
+ * requests in ctx->requests are eligible.
+ */
+static void handle_call_request_backlog (struct sdbus_ctx *ctx)
+{
+    const flux_msg_t *msg;
+    flux_error_t error;
+    msg = flux_msglist_first (ctx->requests);
+    while (msg)  {
+        if (handle_call_request (ctx, msg, &error) < 0) {
+            if (flux_respond_error (ctx->h, msg, errno, error.text) < 0)
+                flux_log_error (ctx->h, "error responding to call request");
+        }
+        msg = flux_msglist_next (ctx->requests);
+    }
+}
+
+/* Bus subscribe completed.  Henceforth, sd-bus signals will be forwarded
+ * to subscribers.  Service pending sdbus.call requests.
+ * N.B. servicing pending sdbus.call requests when the connect is finalized
+ * might cause problems for a user that asynchronously subscribes to signals,
+ * then initiates an action and expects the subscription to capture all signals
+ * triggered by the action.
+ */
+static void subscribe_continuation (flux_future_t *f, void *arg)
```

Comment:
nit: rename to avoid confusion with `subscribe_cb`?  bus_subscribe_continuation?